File: python/refactoring/family_member.py
class FamilyMember:
    def __init__(self, name: str, children: list["FamilyMember"] = []):
        self.name = name
        self.children = children

    def get_descendants(self, generation: int) -> dict["FamilyMember", int]:
        descendants = {}
        for heir in self.children:
            descendants[heir] = generation + 1
            for grand_child in heir.children:
                descendants[grand_child] = generation + 2
                for great_grand_child in grand_child.children:
                    descendants[great_grand_child] = generation + 3
        return descendants

File: csharp/Refactoring/FamilyMember.cs
namespace Refactoring;

public class FamilyMember
{
    public string Name { get; }

    public FamilyMember(string name)
    {
        Name = name;
        this.children = new();
    }

    public FamilyMember(string name, List<FamilyMember> children)
    {
        Name = name;
        this.children = children;
    }

    private List<FamilyMember> children;

    public Dictionary<FamilyMember, int> GetDescendants(int generation)
    {
        Dictionary<FamilyMember, int> descendants = new();
        AddDescendants(children, generation + 1, descendants);
        return descendants;
    }

    private void AddDescendants(IEnumerable<FamilyMember> familyMembers, int generation, Dictionary<FamilyMember, int> descendants)
    {
        foreach (FamilyMember member in familyMembers)
        {
            descendants.Add(member, generation);
            AddDescendants(member.children, generation + 1, descendants);
        }
    }
}
